test/Microsoft.AspNet.Mvc.IntegrationTests/ModelPrefixSelectionIntegrationTest.cs

```diff
@@ -0,0 +1,394 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.Http.Internal;
+using Microsoft.AspNet.Mvc.ModelBinding;
+using Microsoft.Framework.Primitives;
+using Xunit;
+
+namespace Microsoft.AspNet.Mvc.IntegrationTests
+{
+    // Integration tests for the decision logic about how a model-name/prefix is selected at the top-level
+    // of ModelBinding.
+    public class ModelPrefixSelectionIntegrationTest
+    {
+        private class Person1
+        {
+            [FromForm]
+            public string Name { get; set; }
+        }
+
+        [Fact]
+        public async Task ComplexModel_PrefixSelected_ByValueProvider()
+        {
+            // Arrange
+            var argumentBinder = ModelBindingTestHelper.GetArgumentBinder();
+            var parameter = new ParameterDescriptor()
+            {
+                Name = "parameter",
+                ParameterType = typeof(Person1),
+            };
+
+            var operationContext = ModelBindingTestHelper.GetOperationBindingContext(request =>
+            {
+                // This will cause selection of the "parameter" prefix.
+                request.QueryString = new QueryString("?parameter=");
+
+                // This value won't be used, because we select the "parameter" prefix.
+                request.Form = new FormCollection(new Dictionary<string, StringValues>()
+                {
+                    { "Name", "Billy" },
+                });
+            });
+
+            var modelState = new ModelStateDictionary();
+
+            // Act
+            var modelBindingResult = await argumentBinder.BindModelAsync(parameter, modelState, operationContext);
+
+            // Assert
+            Assert.NotNull(modelBindingResult);
+            Assert.True(modelBindingResult.IsModelSet);
+            Assert.Equal("parameter", modelBindingResult.Key);
+
+            var model = Assert.IsType<Person1>(modelBindingResult.Model);
+            Assert.Null(model.Name);
+
+            Assert.Equal(0, modelState.Count);
+            Assert.Equal(0, modelState.ErrorCount);
+            Assert.True(modelState.IsValid);
+        }
+
+        private class Person2
+        {
+            [FromForm]
+            public string Name { get; set; }
+        }
+
+        [Fact]
+        public async Task ComplexModel_PrefixSelected_ByValueProviderValue_WithFilteredValueProviders()
+        {
+            // Arrange
+            var argumentBinder = ModelBindingTestHelper.GetArgumentBinder();
+            var parameter = new ParameterDescriptor()
+            {
+                Name = "parameter",
+                ParameterType = typeof(Person2),
+                BindingInfo = new BindingInfo()
+                {
+                    BindingSource = BindingSource.Query,
```

Comment:
Realize `ComplexModel_EmptyPrefixSelected_NoMatchingValueProviderValue_WithFilteredValueProviders()` is close but has an additional non-empty value provider.